File: neetcode/go/20.valid-parentheses.go
/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */

// @lc code=start
func isValid(s string) bool {
	parenPair := map[rune]rune{
		'(': ')',
		'{': '}',
		'[': ']',
	}

	stack := []rune{}
	for _, paren := range s {
		if _, ok := parenPair[paren]; ok {
			stack = append(stack, paren)
			continue
		}

		last := len(stack) - 1
		if last < 0 || paren != parenPair[stack[last]] {
			return false
		}
		stack = stack[:last]
	}
	return len(stack) == 0
}

// @lc code=end

File: neetcode/python/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        paren_pairs = {
            '(' : ')',
            '{' : '}',
            '[' : ']'
        }

        stack = []
        for paren in s:
            if paren in paren_pairs:
                stack.append(paren)
            elif len(stack) == 0 or paren != paren_pairs[stack.pop()]:
                return False
        return len(stack) == 0 
        
# @lc code=end

